### PR TITLE
fix: Github integration cleanup

### PR DESCRIPTION
**Pull request by adityachoudhari26**

### CODE REVIEW

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubConnectedOrgs.tsx [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  DropdownMenuItem,
+  DropdownMenuTrigger,
+} from "@ctrlplane/ui/dropdown-menu";
+import { Separator } from "@ctrlplane/ui/separator";
+import { Skeleton } from "@ctrlplane/ui/skeleton";
+
+import { api } from "~/trpc/react";
+import { GithubAddOrgDialog } from "./GithubAddOrgDialog";
+
+interface GithubConnectedOrgsProps {
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubConnectedOrgs.tsx:27]: use types for props, stop using cursor

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/page.tsx [hunk 2]
```diff
@@ -43,11 +44,17 @@ export default function GitHubIntegrationPage({
     enabled: session.status === "authenticated",
   });
 
+  const deleteGithubUser = api.github.user.delete.useMutation();
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/page.tsx:47]: fix lint error

**Review by jsbroks — COMMENTED** [review]
fix all the linting issues

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubAddOrgDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                        </Avatar>
+                      )}
+
+                      <span className=" overflow-hidden text-ellipsis">
+                        {value ?? "Select organization..."}
+                      </span>
+                    </div>
+                  </Button>
+                </PopoverTrigger>
+                <PopoverContent className="w-[466px] p-0">
```
Comment (adityachoudhari26) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubAddOrgDialog.tsx:141]: had to explicitly set the width here, using `w-full` or `w-max` does not work

File: apps/webservice/src/app/api/github/installation/route.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    },
+  });
+
+  const baseUrl = await api.runtime.baseUrl();
+  const workspaceSlug = activeWorkspace.slug;
+
+  return NextResponse.redirect(
+    `${baseUrl}/${workspaceSlug}/settings/workspace/integrations/github`,
+  );
+};
```
Comment (zacharyblasczyk) [inline apps/webservice/src/app/api/github/installation/route.ts:162]: This file is cooked. 

```tsx
import type { NextRequest } from "next/server";
import { NextResponse } from "next/server";
import { createAppAuth } from "@octokit/auth-app";
import { Octokit } from "@octokit/rest";
import { BAD_REQUEST, INTERNAL_SERVER_ERROR, NOT_FOUND, UNAUTHORIZED } from "http-status";

import { auth } from "@ctrlplane/auth";
import { eq, takeFirstOrNull } from "@ctrlplane/db";
import { db } from "@ctrlplane/db/client";
import { user, workspace } from "@ctrlplane/db/schema";
import { env } from "~/env";
import { api } from "~/trpc/server";

const createOctokit = (installationId?: number) =>
  env.GITHUB_BOT_APP_ID
    ? new Octokit({
        authStrategy: createAppAuth,
        auth: {
          appId: env.GITHUB_BOT_APP_ID,
          privateKey: env.GITHUB_BOT_PRIVATE_KEY,
          clientId: env.GITHUB_BOT_CLIENT_ID,
          clientSecret: env.GITHUB_BOT_CLIENT_SECRET,
          ...(installationId && { installationId }),
        },
      })
    : null;

const jsonErrorResponse = (message: string, status: number) =>
  NextResponse.json({ error: message }, { status });

export const GET = async (req: NextRequest) => {
  const { searchParams } = new URL(req.url);
  const installationId = searchParams.get("installation_id");
  const setupAction = searchParams.get("setup_action");

  if (!installationId || setupAction !== "install") {
    return jsonErrorResponse("Invalid installation ID or setup action", BAD_REQUEST);
  }

  const session = await auth();
  if (!session) {
    return jsonErrorResponse("Unauthorized", UNAUTHORIZED);
  }

  const userData = await db
    .select()
    .from(user)
    .where(eq(user.id, session.user.id))
    .then(takeFirstOrNull);

  if (!userData?.activeWorkspaceId) {
    return jsonErrorResponse("Workspace not found or unauthorized", NOT_FOUND);
  }

  const activeWorkspace = await db
    .select()
    .from(workspace)
    .where(eq(workspace.id, userData.activeWorkspaceId))
    .then(takeFirstOrNull);

  if (!activeWorkspace) {
    return jsonErrorResponse("Workspace not found", NOT_FOUND);
  }

  const octokit = createOctokit();
  if (!octokit) {
    return jsonErrorResponse("GitHub bot not configured", INTERNAL_SERVER_ERROR);
  }

  const installation = await octokit.apps.getInstallation({
    installation_id: Number(installationId),
    headers: { "X-GitHub-Api-Version": "2022-11-28" },
  });

  if (installation.data.target_type !== "Organization" || !installation.data.account) {
    return jsonErrorResponse("Invalid installation type", BAD_REQUEST);
  }

  const installationOctokit = createOctokit(installation.data.id);
  if (!installationOctokit) {
    return jsonErrorResponse("Failed to get installation octokit", INTERNAL_SERVER_ERROR);
  }

  const orgData = await installationOctokit.orgs.get({
    org: String(installation.data.target_id),
    headers: { "X-GitHub-Api-Version": "2022-11-28" },
  });

  await Promise.all([
    api.github.organizations.create({
      workspaceId: activeWorkspace.id,
      installationId: installation.data.id,
      organizationName: orgData.data.login,
      avatarUrl: orgData.data.avatar_url,
      addedByUserId: userData.id,
    }),
    api.job.agent.create({
      workspaceId: activeWorkspace.id,
      type: "github-app",
      name: orgData.data.login,
      config: {
        installationId: installation.data.id,
        owner: orgData.data.login,
      },
    }),
  ]);

  const baseUrl = await api.runtime.baseUrl();
  return NextResponse.redirect(`${baseUrl}/${activeWorkspace.slug}/settings/workspace/integrations/github`);
};
```

File: apps/webservice/src/app/api/github/installation/route.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+      owner: orgData.data.login,
+    },
+  });
+
+  const baseUrl = await api.runtime.baseUrl();
+  const workspaceSlug = activeWorkspace.slug;
+
+  return NextResponse.redirect(
+    `${baseUrl}/${workspaceSlug}/settings/workspace/integrations/github`,
+  );
```
Comment (zacharyblasczyk) [inline apps/webservice/src/app/api/github/installation/route.ts:161]: I don't think you need the baseurl

File: packages/api/src/router/github/create-github-org.ts [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  new Octokit({
+    authStrategy: createAppAuth,
+    auth: {
+      appId: env.GITHUB_BOT_APP_ID,
+      privateKey: env.GITHUB_BOT_PRIVATE_KEY,
+      clientId: env.GITHUB_BOT_CLIENT_ID,
+      clientSecret: env.GITHUB_BOT_CLIENT_SECRET,
+      installationId,
+    },
+  });
```
Comment (zacharyblasczyk) [inline packages/api/src/router/github/create-github-org.ts:62]: Same here:
```tsx
const createOctokitInstance = (installationId?: number) =>
  new Octokit({
    authStrategy: createAppAuth,
    auth: {
      appId: env.GITHUB_BOT_APP_ID,
      privateKey: env.GITHUB_BOT_PRIVATE_KEY,
      clientId: env.GITHUB_BOT_CLIENT_ID,
      clientSecret: env.GITHUB_BOT_CLIENT_SECRET,
      ...(installationId && { installationId }),
    },
  });
  ```

**Review by zacharyblasczyk — CHANGES_REQUESTED** [review]
I think some things can be simplified.

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubAddOrgDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    userId: string;
+  };
+  children: React.ReactNode;
+  githubConfig: {
+    url: string;
+    botName: string;
+    clientId: string;
+  };
+  workspaceId: string;
+  workspaceSlug: string;
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubAddOrgDialog.tsx:43]: is this used?

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubConnectedOrgs.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+} from "@ctrlplane/ui/card";
+import { Separator } from "@ctrlplane/ui/separator";
+
+import { api } from "~/trpc/server";
+import { GithubAddOrgDialog } from "./GithubAddOrgDialog";
+import { OrgActionDropdown } from "./OrgActionDropdown";
+
+type GithubConnectedOrgsProps = {
+  githubUser?: GithubUser | null;
+  workspaceSlug?: string;
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubConnectedOrgs.tsx:21]: why not just pass in workspace?

workspace: {slug: string, id: string}

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubConnectedOrgs.tsx [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  };
+};
+
+export const GithubConnectedOrgs: React.FC<GithubConnectedOrgsProps> = async ({
+  githubUser,
+  workspaceSlug,
+  workspaceId,
+  githubConfig,
+}) => {
+  const githubOrgsInstalled = await api.github.organizations.list(
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubConnectedOrgs.tsx:37]: do you need to do a null check to enable this query?

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubAddOrgDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+  workspaceId,
+}) => {
+  const [open, setOpen] = useState(false);
+  const [popoverOpen, setPopoverOpen] = useState(false);
+  const githubOrgs = api.github.organizations.byGithubUserId.useQuery(
+    githubUser?.githubUserId ?? 0,
+  );
+  const router = useRouter();
+  const githubOrgsInstalled =
+    api.github.organizations.list.useQuery(workspaceId);
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubAddOrgDialog.tsx:59]: do you need to do a null check to enable this query?

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubAddOrgDialog.tsx [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+export const GithubAddOrgDialog: React.FC<GithubAddOrgDialogProps> = ({
+  githubUser,
+  children,
+  githubConfig,
+  workspaceId,
+}) => {
+  const [open, setOpen] = useState(false);
+  const [popoverOpen, setPopoverOpen] = useState(false);
+  const githubOrgs = api.github.organizations.byGithubUserId.useQuery(
+    githubUser?.githubUserId ?? 0,
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubAddOrgDialog.tsx:55]: do you need to do a null check to enable this query?

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubRemoveOrgDialog.tsx [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+import { api } from "~/trpc/react";
+
+interface GithubConfigFileWithDeployments extends GithubConfigFile {
+  deployments: Deployment[];
+}
+
+export interface GithubOrganizationWithConfigFiles extends GithubOrganization {
+  configFiles: GithubConfigFileWithDeployments[];
+}
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubRemoveOrgDialog.tsx:30]: use types, and use `&` instead, this is the OOP

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubAddOrgDialog.tsx [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+          <>
+            <Separator className="my-4" />
+            <div className="flex flex-col gap-4 px-4">
+              <DialogHeader>
+                <DialogTitle>
+                  Select from pre-connected organizations
+                </DialogTitle>
+                <DialogDescription>
+                  <div
+                    className="relative mb-4 flex w-fit flex-col gap-2 rounded-md bg-neutral-800/50 px-4 py-3 text-muted-foreground"
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/GithubAddOrgDialog.tsx:120]: flex row, aling icon at top of row

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/OrgActionDropdown.tsx [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+          <div className="h-2 w-2 rounded-full bg-green-500" />
+          Connected
+          <TbChevronDown className="h-4 w-4" />
+        </Button>
+      </DropdownMenuTrigger>
+      <DropdownMenuContent>
+        <Link
+          href={`${githubConfig.url}/organizations/${org.organizationName}/settings/installations/${org.installationId}`}
+          target="_blank"
+          rel="noopener noreferrer"
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/OrgActionDropdown.tsx:43]: lmfao bet you have no idea what cursor is doing here

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/page.tsx [hunk 15]
```diff
@@ -20,33 +21,28 @@ const githubAuthUrl = (
...
+  const workspace = await api.workspace.bySlug(workspaceSlug);
+  if (workspace == null) return notFound();
+  const session = await auth();
+  if (session == null) redirect("/login");
+
+  const baseUrl = env.BASE_URL;
 
-  const githubUrl = api.runtime.github.url.useQuery();
-  const githubBotName = api.runtime.github.botName.useQuery();
-  const githubBotClientId = api.runtime.github.clientId.useQuery();
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/page.tsx:36]: can we remove these endpoints

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/page.tsx [hunk 16]
```diff
@@ -61,57 +57,54 @@ export default function GitHubIntegrationPage({
...
-            clientId: githubBotClientId.data ?? "",
-          }}
-        />
-      )}
+        {isGithubConfigured && (
+          <GithubConnectedOrgs
+            githubUser={githubUser}
+            workspaceId={workspace.id}
+            workspaceSlug={workspaceSlug}
+            loading={false}
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/workspace/integrations/(integration)/github/page.tsx:97]: remove loading...

File: packages/api/src/router/github/create-github-org.ts [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+  workspace,
+} from "@ctrlplane/db/schema";
+import { configFile } from "@ctrlplane/validators";
+
+import { env } from "../../config";
+
+type ConfigFile =
+  RestEndpointMethodTypes["search"]["code"]["response"]["data"]["items"][number];
+
+type ParsedConfigFile = ConfigFile & {
```
Comment (jsbroks) [inline packages/api/src/router/github/create-github-org.ts:27]: zod supports `z.infer<typeof configFile>`

File: packages/api/src/router/github/create-github-org.ts [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+    .then(({ data }) => {
+      if (!("content" in data)) throw new Error("Invalid response data");
+      return Buffer.from(data.content, "base64").toString("utf-8");
+    });
+
+  const yamlContent = yaml.load(content);
+  const parsed = configFile.safeParse(yamlContent);
+  if (!parsed.success) throw new Error("Invalid config file");
+
+  return {
```
Comment (jsbroks) [inline packages/api/src/router/github/create-github-org.ts:86]: 1 line

File: packages/api/src/router/github/create-github-org.ts [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+  const deploymentInfo = await db
+    .select()
+    .from(system)
+    .innerJoin(workspace, eq(system.workspaceId, workspace.id))
+    .where(
+      and(
+        inArray(
+          system.slug,
+          parsedConfigFiles
+            .map((d) => d.content.deployments.map((d) => d.system))
```
Comment (jsbroks) [inline packages/api/src/router/github/create-github-org.ts:106]: flatMap

File: packages/api/src/router/github/create-github-org.ts [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+            .flat(),
+        ),
+        inArray(
+          workspace.slug,
+          parsedConfigFiles
+            .map((d) => d.content.deployments.map((d) => d.workspace))
```
Comment (jsbroks) [inline packages/api/src/router/github/create-github-org.ts:112]: flatMap

File: packages/api/src/router/github/create-github-org.ts [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+        q: `org:${org.organizationName} filename:ctrlplane.yml`,
+        per_page: 100,
+        headers: {
+          "X-GitHub-Api-Version": "2022-11-28",
+          authorization: `Bearer ${installationToken.token}`,
+        },
+      }),
+    ]).then((responses) => {
+      return [...responses[0].data.items, ...responses[1].data.items];
+    });
```
Comment (jsbroks) [inline packages/api/src/router/github/create-github-org.ts:205]: 1 line, no need for return

be more descriptive, give these vars names for the data they are providing

```suggestion
    ]).then(([valueA, valueB]) => 
      [...valueA.data.items, ...valueBdata.items]
    );
```

File: packages/api/src/router/github/github.ts [hunk 22]
```diff
@@ -0,0 +1,349 @@
...
+      ctx.db
+        .select()
+        .from(githubUser)
+        .where(eq(githubUser.userId, input))
+        .then(takeFirstOrNull),
+    ),
+
+  delete: protectedProcedure
+    .input(z.string())
+    .mutation(async ({ ctx, input }) =>
```
Comment (jsbroks) [inline packages/api/src/router/github/github.ts:69]: need aysnc here?

File: packages/api/src/router/github/github.ts [hunk 23]
```diff
@@ -0,0 +1,349 @@
...
+        githubUserId: z.number(),
+        githubUsername: z.string(),
+      }),
+    )
+    .mutation(({ ctx, input }) =>
+      ctx.db
+        .insert(githubUser)
+        .values(input)
+        .returning()
+        .then((data) => data[0]),
```
Comment (jsbroks) [inline packages/api/src/router/github/github.ts:86]:
```suggestion
        .then(([data]) => data),
```

why do we need this? use takeFirst?

**Pull request closed without merging**